contrib/terraform-testing-infrastructure/modules/config-files/templates/initialize_hadoop.sh.tftpl

```diff
@@ -28,9 +28,15 @@ pdsh -S -g worker sudo systemctl enable hadoop-datanode yarn-nodemanager
 # 3. Start the namenode
 # 4. Start the resource manager
 # 5. Start datanodes and nodemanagers on the worker nodes
+# 6. Wait for the namenode to leave safemode
 #
 sudo systemctl start zookeeper
 hdfs namenode -format
 sudo systemctl start hadoop-namenode
 sudo systemctl start yarn-resourcemanager
 pdsh -S -g worker 'sudo systemctl start hadoop-datanode yarn-nodemanager'
+echo "Waiting for NameNode to leave safemode..."
+while [[ ! $(hdfs dfsadmin -safemode wait) ]]; do
```

Comment:
Looking at the [source code](https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java) for this, it would seem that this command waits internally and always returns success. It will print out whether the namenode is actually in safe mode or not, so it would seem that's how you have to check if it failed, but maybe I misread the code.

Reply:
Are you suggesting that `waitExitSafeMode` does not wait?

https://github.com/apache/hadoop/blob/trunk/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/DFSAdmin.java#L688

Reply:
Alternatively:

```suggestion
echo "Waiting for NameNode to leave safemode..."
while [[ ! $(hdfs dfsadmin -safemode wait 2>/dev/null) =~ OFF ]]; do
  sleep 5
  echo "Waiting for NameNode to leave safemode..."
done
```